### PR TITLE
Handling max/min values in CommandLineArgumentParser

### PR DESCRIPTION
**Pull request by magicDGS**
### Included
* Handling max/min values in `CommandLineArgumentParser` (solves partially #6)
* New `CommandLineArgumentException` for out-of-range values.

### Not included
* `LegacyCommandLineArgumentParser` does not have max/min validation: Picard is not supporting currently this feature in its parser.
* Recommended boundaries are not handled yet.

### CODE REVIEW

**Comment by magicDGS** [comment]
This is very simple but useful for my work. Could you have a look to this, @cmnbroad?

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineException.java [hunk 1]
```diff
@@ -39,6 +39,28 @@ public BadArgumentValue(String message) {
...
+
+        public OutOfRangeArgumentValue(final String arg, final double minValue, final double maxValue, final double value) {
+            super(arg, Double.toString(value), getMessage(minValue, maxValue));
+        }
+
+        private static String getMessage(final double minValue, final double maxValue) {
+            final boolean outMinValue = minValue != Double.NEGATIVE_INFINITY;
+            final boolean outMaxValue = maxValue != Double.POSITIVE_INFINITY;
+            if (outMinValue && outMaxValue) {
+                return String.format("allowed range (%.2f, %.2f).", minValue, maxValue);
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineException.java]: I would widen the precision when converting these to strings, otherwise the values may be unrecognizable. Maybe .5f is reasonable ? Or maybe Double.toString will do a better job ?

File: src/main/java/org/broadinstitute/barclay/argparser/LegacyCommandLineArgumentParser.java [hunk 2]
```diff
@@ -548,7 +548,8 @@ private boolean parseOption(String key, final String stringValue, final boolean
             } else {
                 value = constructFromString(getUnderlyingType(optionDefinition.field), stringValue);
             }
-
+            // TODO - Picard legacy parsing is not taking into account range within options
+            // TODO - should we add support here?
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/argparser/LegacyCommandLineArgumentParser.java]: I don't think we want to invest in adding support for these to the legacy parser. Perhaps just add some code to log a warning saying its unsupported and not enforced if a value is specified. (We don't want to throw since the new parser will handle them, and we want to be able to switch back and forth between the two parsers).

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java [hunk 3]
```diff
@@ -498,6 +498,26 @@ private void validatePluginArguments() {
...
+            return;
+        }
+        final double argumentDoubleValue = ((Number)argumentValue).doubleValue();
+
+        // Check hard limits first, if specified
+        if ((argumentDefinition.minValue != Double.NEGATIVE_INFINITY && argumentDoubleValue < argumentDefinition.minValue)
+            || (argumentDefinition.maxValue != Double.POSITIVE_INFINITY && argumentDoubleValue > argumentDefinition.maxValue)) {
+            throw new CommandLineException.OutOfRangeArgumentValue(argumentDefinition.fullName, argumentDefinition.minValue, argumentDefinition.maxValue, argumentDoubleValue);
+        }
+    }
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: While we're doing this, we might want to log a warning if the value is outside the minRec/maxRec range as well (you didn't include those annotation fields in this PR, but when you rebase this they'll be there!).

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java [hunk 4]
```diff
@@ -498,6 +498,26 @@ private void validatePluginArguments() {
...
     }
+    /**
+     * Check the provided value against any range constraints specified in the Argument annotation
+     * for the corresponding field. Throw an exception if limits are violated.
+     *
+     * - Only checks numeric types (int, double, etc.)
+     */
+    private void checkArgumentRange(final ArgumentDefinition argumentDefinition, final Object argumentValue) {
+        // Only validate numeric types
+        if ( ! (argumentValue instanceof Number) ) {
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: argumentValue can be null if the user specifies "null". This code won't issue a validation failure for that case, but it probably should.

File: src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java [hunk 5]
```diff
@@ -749,6 +749,56 @@ private static String captureSystemStream(Runnable runnable, PrintStream stream,
...
+            summary = "tool with boundaries",
+            oneLineSummary = "tools with boundaries",
+            programGroup = TestProgramGroup.class
+    )
+    public class WithBoundariesArguments {
+
+        @Argument(doc = "Integer in the range [0, 30]", optional = true, minValue = 0, maxValue = 30)
+        public int integerArg = 20;
+
+        @Argument(doc = "String argument", optional = true, minValue = 0, maxValue = 30)
```
Comment (cmnbroad) [inline src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java]: Per my comment above, I think this case should throw, since the range is not honored/enforced. We should keep a test for this though to make sure it fails as we think it should.

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineException.java [hunk 6]
```diff
@@ -39,6 +39,28 @@ public BadArgumentValue(String message) {
...
+            } else if (outMinValue) {
+                return String.format("minimum allowed value %.2f", minValue);
+            } else if (outMaxValue) {
+                return String.format("maximum allowed value %.2f", maxValue);
+            }
+            return "unkwnon";
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineException.java]: Spelling. "unknown"

File: src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java [hunk 7]
```diff
@@ -749,6 +749,56 @@ private static String captureSystemStream(Runnable runnable, PrintStream stream,
...
+        Assert.assertEquals(o.stringArg, expectedString);
+    }
+
+    @DataProvider
+    public Object[][] outOfRangeArgs() {
+        return new Object[][]{
+                {new String[]{"--integerArg", "-45"}},
+                {new String[]{"--integerArg", "-1"}},
+                {new String[]{"--integerArg", "31"}},
+                {new String[]{"--integerArg", "106"}}
```
Comment (cmnbroad) [inline src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java]: Can you add a test that uses the value "null". Thx.

File: src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java [hunk 8]
```diff
@@ -749,6 +749,56 @@ private static String captureSystemStream(Runnable runnable, PrintStream stream,
...
 
+    @CommandLineProgramProperties(
+            summary = "tool with boundaries",
+            oneLineSummary = "tools with boundaries",
+            programGroup = TestProgramGroup.class
+    )
+    public class WithBoundariesArguments {
+
+        @Argument(doc = "Integer in the range [0, 30]", optional = true, minValue = 0, maxValue = 30)
+        public int integerArg = 20;
```
Comment (cmnbroad) [inline src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java]: Add a double argument test as well.

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineException.java (hunk 1 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineException.java]: It would also be nice to handle the integer valued case correctly in the output strings (and not print out floating point values for an integer argument).

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java [hunk 9]
```diff
@@ -988,6 +1013,9 @@ public ArgumentDefinition(
             //null collections have been initialized by createCollection which is called in handleArgumentAnnotation
             //this is optional if it's specified as being optional or if there is a default value specified
             this.optional = annotation.optional() || ! this.defaultValue.equals(NULL_STRING);
+
+            this.maxValue = annotation.maxValue();
+            this.minValue = annotation.minValue();
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: Also, it might be worth checking that the author hasn't provided a non-integral value for an integer arg.

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java (hunk 4 above)
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: I added validation for null values, and here it only returns if the value is not numeric and not null.

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java (hunk 3 above)
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: I rebased, added a logger in the class and log a warning if the value is out of the soft boundaries.

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java (hunk 9 above)
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: I check both on `ArgumentDefinition`construction, for both hard and soft boundaries.

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineException.java [hunk 10]
```diff
@@ -39,6 +39,28 @@ public BadArgumentValue(String message) {
         }
     }
 
+    public static class OutOfRangeArgumentValue extends BadArgumentValue {
+        private static final long serialVersionUID = 0L;
+
+        public OutOfRangeArgumentValue(final String arg, final double minValue, final double maxValue, final double value) {
```
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineException.java]: Done.

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineException.java (hunk 1 above)
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineException.java]: I changed the signature of the constructor for take an `Object` and check if it is an `Integer` instance. Now for double it uses `Double.toString(value)` and `Integer.toString((int) Math.rint(value)`.

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineException.java (hunk 6 above)
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineException.java]: Changed, although this point should never be reached. Maybe is better to thrown an `IllegalArgumentException` if so?

File: src/main/java/org/broadinstitute/barclay/argparser/LegacyCommandLineArgumentParser.java (hunk 2 above)
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/LegacyCommandLineArgumentParser.java]: Should we log the warning on `OptionDefinition` construction?

File: src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java (hunk 8 above)
Comment (magicDGS) [inline src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java]: Done.

File: src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java (hunk 5 above)
Comment (magicDGS) [inline src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java]: Added test for throwing `CommandLineException.CommandLineParserInternalException`, and removed boundaries here.

File: src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java (hunk 7 above)
Comment (magicDGS) [inline src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java]: Added.

**Comment by magicDGS** [comment]
I did some workarround for the implementation to handle everything in a better way. Thanks for reviewing @cmnbroad!

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java [hunk 11]
```diff
@@ -947,6 +1000,7 @@ private Object constructFromString(final Class clazz, final String s, final Stri
 
     public static class ArgumentDefinition {
         public final Field field;
+        public final Class<?> type;
```
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: This could be used in other parts of the code in this class instead of calling `CommandLineParser.getUnderlyingType()`. This will simplify the code a bit. Should I change it here, @cmnbroad?

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java [hunk 12]
```diff
@@ -998,9 +1056,31 @@ public ArgumentDefinition(
...
+                        || this.minRecommendedValue != Double.NEGATIVE_INFINITY) {
+                    throw new CommandLineException.CommandLineParserInternalException("Bounds for arguments could only be set for numeric arguments");
+                }
+            }
+            if (Integer.class.isAssignableFrom(this.type)) {
+                if (!isInfinityOrMathematicalInteger(this.maxValue)
+                        || !isInfinityOrMathematicalInteger(this.minValue)
+                        || !isInfinityOrMathematicalInteger(this.maxRecommendedValue)
+                        || !isInfinityOrMathematicalInteger(this.minRecommendedValue)) {
+                    throw new CommandLineException.CommandLineParserInternalException("Bounds for integer arguments should be mathematical integers");
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: Same thing for this; test of the message should include the offending field name.

Also, please add a test for this code path.

File: src/main/java/org/broadinstitute/barclay/argparser/LegacyCommandLineArgumentParser.java (hunk 2 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/argparser/LegacyCommandLineArgumentParser.java]: Yes, sounds like a reasonable place (please include the name of the argument int he message here as well).

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java [hunk 13]
```diff
@@ -508,6 +513,51 @@ private void validatePluginArguments() {
...
+        // Check hard limits first, if specified
+        if (isOutOfRange(argumentDefinition.minValue, argumentDefinition.maxValue, argumentDoubleValue)) {
+            throw new CommandLineException.OutOfRangeArgumentValue(argumentDefinition.fullName, argumentDefinition.minValue, argumentDefinition.maxValue, argumentValue);
+        }
+        // Check recommended values
+        if (isOutOfRange(argumentDefinition.minRecommendedValue, argumentDefinition.maxRecommendedValue, argumentDoubleValue)) {
+            final boolean outMinValue = argumentDefinition.minRecommendedValue != Double.NEGATIVE_INFINITY;
+            final boolean outMaxValue = argumentDefinition.maxRecommendedValue != Double.POSITIVE_INFINITY;
+            if (outMinValue && outMaxValue) {
+                logger.warn("Argument --{} has value {}, but recommended within range ({},{})",
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: We should have test cases that exercise these code paths (I don't think there are any ?), especially if we're going to use the {} style formatting, to at least ensure that we don't blow up here.

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java [hunk 14]
```diff
@@ -508,6 +513,51 @@ private void validatePluginArguments() {
...
+                logger.warn("Argument --{} has value {}, but minimum recommended is {}",
+                        argumentDefinition.fullName, argumentDoubleValue, argumentDefinition.minRecommendedValue);
+            } else if (outMaxValue) {
+                logger.warn("Argument --{} has value {}, but maximum recommended is {}",
+                        argumentDefinition.fullName, argumentDoubleValue, argumentDefinition.maxRecommendedValue);
+            }
+        }
+    }
+
+    private static boolean isOutOfRange(final double minValue, final double maxValue, final Double value) {
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: Add a comment here saying we consider a null value to be out of range.

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java [hunk 15]
```diff
@@ -508,6 +513,51 @@ private void validatePluginArguments() {
...
         pluginDescriptors.entrySet().forEach(e -> e.getValue().validateArguments());
     }
+    /**
+     * Check the provided value against any range constraints specified in the Argument annotation
+     * for the corresponding field. Throw an exception if limits are violated.
+     *
+     * - Only checks numeric types (int, double, etc.)
+     */
+    private void checkArgumentRange(final ArgumentDefinition argumentDefinition, final Object argumentValue) {
+        // Only validate numeric types
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: Please add a comment here saying that we only check numeric types because we have already ensured that only numeric types have bounds.

File: src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java [hunk 16]
```diff
@@ -749,6 +749,82 @@ private static String captureSystemStream(Runnable runnable, PrintStream stream,
...
+    )
+    public class WithBoundariesArguments {
+
+        @Argument(doc = "Double argument in the range [10, 20]", optional = true, minValue = 10, maxValue = 20)
+        public Double doubleArg = 15d;
+
+        @Argument(doc = "Integer in the range [0, 30]", optional = true, minValue = 0, maxValue = 30)
+        public int integerArg = 20;
+
+        @Argument(doc = "String argument", optional = true)
```
Comment (cmnbroad) [inline src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java]: This field seems redundant and out of place now with the most recent changes ?

File: src/main/java/org/broadinstitute/barclay/argparser/LegacyCommandLineArgumentParser.java (hunk 2 above)
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/LegacyCommandLineArgumentParser.java]: Done before the construction of `OptionDefinition` in `handleOptionAnnotation()`. Removed TODO here.

File: src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java [hunk 17]
```diff
@@ -749,6 +749,82 @@ private static String captureSystemStream(Runnable runnable, PrintStream stream,
...
+    public void testWithBoundariesArgumentsForNoNumeric() {
+        @CommandLineProgramProperties(summary = "broken tool",
+                oneLineSummary = "broken tool",
+                programGroup = TestProgramGroup.class)
+        class WithBoundariesArgumentsForString {
+            @Argument(doc = "String argument", optional = true, minValue = 0, maxValue = 30)
+            public String stringArg = "string";
+        }
+        final CommandLineArgumentParser clp = new CommandLineArgumentParser(new WithBoundariesArgumentsForString());
+        clp.parseArguments(System.err, new String[]{});
```
Comment (magicDGS) [inline src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java]: Done

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java (hunk 15 above)
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: Done.

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java (hunk 13 above)
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: Added recommended values to the current test.

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java [hunk 18]
```diff
@@ -998,9 +1056,31 @@ public ArgumentDefinition(
...
+            this.minRecommendedValue = annotation.minRecommendedValue();
+            // bounds should be only set for numeric arguments and if the type is integer it should
+            // be set to an integer
+            this.type = CommandLineParser.getUnderlyingType(this.field);
+            if (! Number.class.isAssignableFrom(this.type)) {
+                if (this.maxValue != Double.POSITIVE_INFINITY
+                        || this.minValue != Double.NEGATIVE_INFINITY
+                        || this.maxRecommendedValue != Double.POSITIVE_INFINITY
+                        || this.minRecommendedValue != Double.NEGATIVE_INFINITY) {
+                    throw new CommandLineException.CommandLineParserInternalException("Bounds for arguments could only be set for numeric arguments");
```
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: Done

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java (hunk 12 above)
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: Done (both)

File: src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java (hunk 16 above)
Comment (magicDGS) [inline src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java]: I maintain it for checking if it does not blows up if the argument is a `String`, but it's true that it is redundant with other tests already implemented. Removed!

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java [hunk 19]
```diff
@@ -508,6 +513,52 @@ private void validatePluginArguments() {
...
+        // Only validate numeric types because we have already ensured at constructor time that only numeric types have bounds
+        if (!Number.class.isAssignableFrom(argumentDefinition.type)) {
+            return;
+        }
+
+        final Double argumentDoubleValue = (argumentValue == null) ? null : ((Number)argumentValue).doubleValue();
+
+        // Check hard limits first, if specified
+        if (isOutOfRange(argumentDefinition.minValue, argumentDefinition.maxValue, argumentDoubleValue)) {
+            throw new CommandLineException.OutOfRangeArgumentValue(argumentDefinition.fullName, argumentDefinition.minValue, argumentDefinition.maxValue, argumentValue);
```
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: I realized that I'm using here the `argumentDefinition.fullName` field, but I guess that it is better to use the `getLongName()`, no?

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java [hunk 20]
```diff
@@ -508,6 +513,52 @@ private void validatePluginArguments() {
...
+        }
+        // Check recommended values
+        if (isOutOfRange(argumentDefinition.minRecommendedValue, argumentDefinition.maxRecommendedValue, argumentDoubleValue)) {
+            final boolean outMinValue = argumentDefinition.minRecommendedValue != Double.NEGATIVE_INFINITY;
+            final boolean outMaxValue = argumentDefinition.maxRecommendedValue != Double.POSITIVE_INFINITY;
+            if (outMinValue && outMaxValue) {
+                logger.warn("Argument --{} has value {}, but recommended within range ({},{})",
+                        argumentDefinition.fullName, argumentDoubleValue, argumentDefinition.minRecommendedValue, argumentDefinition.maxRecommendedValue);
```
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: The same here and other warnings.

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java [hunk 21]
```diff
@@ -1068,15 +1069,15 @@ public ArgumentDefinition(
...
+                    throw new CommandLineException.CommandLineParserInternalException(String.format("Min/max value ranges can only be set for numeric arguments. Argument --%s has a minimum or maximum value but has a non-numeric type.", this.getLongName()));
                 }
             }
             if (Integer.class.isAssignableFrom(this.type)) {
                 if (!isInfinityOrMathematicalInteger(this.maxValue)
                         || !isInfinityOrMathematicalInteger(this.minValue)
                         || !isInfinityOrMathematicalInteger(this.maxRecommendedValue)
                         || !isInfinityOrMathematicalInteger(this.minRecommendedValue)) {
-                    throw new CommandLineException.CommandLineParserInternalException("Bounds for integer arguments should be mathematical integers");
+                    throw new CommandLineException.CommandLineParserInternalException(String.format("Integer argument --%s has a minimum or maximum value but has a non-integer type.", this.getLongName()));
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: How  about "has a minimum or maximum value but has a non-integer type" -> "has a minimum or maximum attribute with a non-integral value"

File: src/main/java/org/broadinstitute/barclay/argparser/LegacyCommandLineArgumentParser.java [hunk 22]
```diff
@@ -807,6 +809,19 @@ private void handleOptionAnnotation(final Field field, final int fieldPosition)
                 printOrder = fieldPosition * 1000;
             }
 
+            // log a warning if boundaries are set
+            if (optionAnnotation.maxValue() != Double.POSITIVE_INFINITY) {
+                logger.warn("UNSUPPORTED: maximum allowed value for argument --{} is not forced", field.getName());
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/argparser/LegacyCommandLineArgumentParser.java]: Thanks for adding these. I would remove the phrase "UNSUPPORTED: "  from all four of these, since we really are just issuing a warning. Also, I think these messages should use the argument name, not the field name (sorry if I misled you in the last round of comments). Finally, "forced" should -> "enforced" in these strings.

File: src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java [hunk 23]
```diff
@@ -769,39 +780,34 @@ public void testWithBoundariesArgumentsForNoNumeric() {
     )
     public class WithBoundariesArguments {
 
-        @Argument(doc = "Double argument in the range [10, 20]", optional = true, minValue = 10, maxValue = 20)
+        @Argument(doc = "Double argument in the range [10, 20]", optional = true, minValue = 10, minRecommendedValue = 16, maxRecommendedValue = 17, maxValue = 20)
```
Comment (cmnbroad) [inline src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java]: Can you add a comment to these saying that the recommended values here are not explicitly verified by the tests, but do force the code through the warning code paths.

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java (hunk 19 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: Yes, that would be better.

Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: Done

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java (hunk 20 above)
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: Done.

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java (hunk 21 above)
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java]: Done.

File: src/main/java/org/broadinstitute/barclay/argparser/LegacyCommandLineArgumentParser.java (hunk 22 above)
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/LegacyCommandLineArgumentParser.java]: I removed the phrase "UNSUPPORTED" and use "enforced", but I think that the argument name is the field name for the legacy parser. Am I wrong?

Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/LegacyCommandLineArgumentParser.java]: I moved this warnings down and use `OptionDefinition.name` to be clearer here. I guess that's what you mean.

File: src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java (hunk 23 above)
Comment (magicDGS) [inline src/test/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParserTest.java]: Done.

**Comment by magicDGS** [comment]
I addressed the latest comments. Back to you @cmnbroad and many thanks for reviewing!

File: src/main/java/org/broadinstitute/barclay/argparser/CommandLineException.java [hunk 24]
```diff
@@ -39,6 +41,36 @@ public BadArgumentValue(String message) {
...
+            final DoubleFunction<String> toString = (asInt) ? Double::toString : v -> Integer.toString((int) Math.rint(v));
+            if (outMinValue && outMaxValue) {
+                return String.format("allowed range [%s, %s].", toString.apply(minValue), toString.apply(maxValue));
+            } else if (outMinValue) {
+                return String.format("minimum allowed value %s", toString.apply(minValue));
+            } else if (outMaxValue) {
+                return String.format("maximum allowed value %s", toString.apply(maxValue));
+            }
+            // this should never be reached
+            return "unknown";
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineException.java]: @magicDGS I know you asked about this in the last round but I never commented. I think we should throw here since it shouldn't ever happen.

Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/argparser/CommandLineException.java]: Done

**Comment by magicDGS** [comment]
Sorry for the delay, I was on holidays. I squashed/rebased my commits and included the exception thrown. Back to you,  @cmnbroad, and thanks for reviewing.

**Comment by magicDGS** [comment]
Thanks for this, @cmnbroad!
